File: app/src/main/java/com/intkhabahmed/smartnotes/adapters/NotesAdapter.java
package com.intkhabahmed.smartnotes.adapters;

import android.content.Context;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.intkhabahmed.smartnotes.R;
import com.intkhabahmed.smartnotes.models.ChecklistItem;
import com.intkhabahmed.smartnotes.models.Note;
import com.intkhabahmed.smartnotes.utils.NoteUtils;

import java.io.File;
import java.util.List;

/**
 * Created by INTKHAB on 23-03-2018.
 */

public class NotesAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final Context mContext;
    private List<Note> mNotes;
    private OnItemClickListener mOnItemClickListener;

    public NotesAdapter(Context context, OnItemClickListener clickListener) {
        mContext = context;
        mOnItemClickListener = clickListener;
    }

    public interface OnItemClickListener {
        void onMenuItemClick(View view, Note note);

        void onItemClick(int noteId, String noteType);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view;
        switch (viewType) {
            case 0:
                view = LayoutInflater.from(mContext).inflate(R.layout.note_item, parent, false);
                return new TextNotesViewHolder(view);
            case 1:
                view = LayoutInflater.from(mContext).inflate(R.layout.image_note_item, parent, false);
                return new ImageNotesViewHolder(view);
            default:
                throw new IllegalArgumentException("Unknown ViewType");
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Note note = mNotes.get(holder.getAdapterPosition());
        switch (holder.getItemViewType()) {
            case 0:
                TextNotesViewHolder textNotesViewHolder = (TextNotesViewHolder) holder;
                if (note.getNoteType().equals(mContext.getString(R.string.simple_note))) {
                    textNotesViewHolder.noteDescriptionTextView.setText(note.getDescription());
                } else if (note.getNoteType().equals(mContext.getString(R.string.checklist))) {
                    List<ChecklistItem> checklistItems = new Gson().fromJson(note.getDescription(), new TypeToken<List<ChecklistItem>>() {
                    }.getType());
                    int noOfItems = checklistItems.size() >= 2 ? 2 : checklistItems.size();
                    textNotesViewHolder.noteDescriptionTextView.setText("");
                    for (int i = 0; i < noOfItems; i++) {
                        textNotesViewHolder.noteDescriptionTextView
                                .append(String.valueOf(checklistItems.get(i).getTitle()));
                        if (checklistItems.get(i).isChecked()) {
                            textNotesViewHolder.noteDescriptionTextView.append(" " +
                                    mContext.getString(R.string.checkmark_unicode));
                        }
                        if (i < noOfItems - 1) {
                            textNotesViewHolder.noteDescriptionTextView.append("\n");
                        }
                    }
                    if (checklistItems.size() > 2) {
                        textNotesViewHolder.noteDescriptionTextView.append(" ...");
                    }
                }
                textNotesViewHolder.noteTitleTextView.setText(note.getNoteTitle());
                textNotesViewHolder.noteCreateDateTextView.setText(NoteUtils.getFormattedTime(note.getDateCreated(),
                        System.currentTimeMillis()));
                break;
            case 1:
                ImageNotesViewHolder imageNotesViewHolder = (ImageNotesViewHolder) holder;
                File imageFile = new File(note.getDescription());
                if (imageFile.exists()) {
                    Glide.with(mContext).asDrawable().load(Uri.fromFile(imageFile)).into(imageNotesViewHolder.noteImageView);
                }
                imageNotesViewHolder.noteTitleTextView.setText(note.getNoteTitle());
                imageNotesViewHolder.noteCreateDateTextView.setText(NoteUtils.getFormattedTime(note.getDateCreated(),
                        System.currentTimeMillis()));
                break;
        }
    }

    @Override
    public int getItemCount() {
        if (mNotes == null) {
            return 0;
        }
        return mNotes.size();
    }

    @Override
    public int getItemViewType(int position) {
        Note note = mNotes.get(position);
        if (note.getNoteType().equals(mContext.getString(R.string.simple_note))
                || note.getNoteType().equals(mContext.getString(R.string.checklist))) {
            return 0;
        }
        return 1;
    }

    public void setNotes(List<Note> notes) {
        mNotes = notes;
        notifyDataSetChanged();
    }

    public class TextNotesViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView noteTitleTextView;
        TextView noteDescriptionTextView;
        TextView noteCreateDateTextView;
        ImageButton itemMenuButton;

        private TextNotesViewHolder(View itemView) {
            super(itemView);
            noteTitleTextView = itemView.findViewById(R.id.note_title);
            noteDescriptionTextView = itemView.findViewById(R.id.note_description);
            noteCreateDateTextView = itemView.findViewById(R.id.note_create_date);
            itemMenuButton = itemView.findViewById(R.id.item_menu_button);
            noteTitleTextView.setOnClickListener(this);
            noteDescriptionTextView.setOnClickListener(this);
            noteCreateDateTextView.setOnClickListener(this);
            itemMenuButton.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (view instanceof ImageButton) {
                mOnItemClickListener.onMenuItemClick(view, mNotes.get(getAdapterPosition()));
            } else if (view instanceof TextView) {
                mOnItemClickListener.onItemClick(mNotes.get(getAdapterPosition()).getNoteId(), mNotes.get(getAdapterPosition()).getNoteType());
            }
        }
    }

    public class ImageNotesViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView noteTitleTextView;
        ImageView noteImageView;
        TextView noteCreateDateTextView;
        ImageButton itemMenuButton;

        private ImageNotesViewHolder(View itemView) {
            super(itemView);
            noteTitleTextView = itemView.findViewById(R.id.tv_note_title);
            noteImageView = itemView.findViewById(R.id.note_image_view);
            noteCreateDateTextView = itemView.findViewById(R.id.tv_note_date_created);
            itemMenuButton = itemView.findViewById(R.id.note_menu_button);
            noteTitleTextView.setOnClickListener(this);
            noteImageView.setOnClickListener(this);
            noteCreateDateTextView.setOnClickListener(this);
            itemMenuButton.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (view instanceof ImageButton) {
                mOnItemClickListener.onMenuItemClick(view, mNotes.get(getAdapterPosition()));
            } else if (view instanceof TextView || view instanceof ImageView) {
                mOnItemClickListener.onItemClick(mNotes.get(getAdapterPosition()).getNoteId(), mNotes.get(getAdapterPosition()).getNoteType());
            }
        }
    }
}

File: app/src/main/java/com/intkhabahmed/smartnotes/adapters/NotesAdapter.kt
package com.intkhabahmed.smartnotes.adapters

import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import com.bumptech.glide.Glide
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.intkhabahmed.smartnotes.R
import com.intkhabahmed.smartnotes.adapters.view.BaseViewHolder
import com.intkhabahmed.smartnotes.models.ChecklistItem
import com.intkhabahmed.smartnotes.models.Note
import com.intkhabahmed.smartnotes.utils.AppConstants
import com.intkhabahmed.smartnotes.utils.NoteUtils
import kotlinx.android.synthetic.main.image_note_item.view.note_image_view
import kotlinx.android.synthetic.main.image_note_item.view.note_menu_button
import kotlinx.android.synthetic.main.image_note_item.view.tv_note_date_created
import kotlinx.android.synthetic.main.image_note_item.view.tv_note_title
import kotlinx.android.synthetic.main.note_item.view.item_menu_button
import kotlinx.android.synthetic.main.note_item.view.note_create_date
import kotlinx.android.synthetic.main.note_item.view.note_description
import kotlinx.android.synthetic.main.note_item.view.note_title
import java.io.File

/**
 * Created by INTKHAB on 23-03-2018.
 */

class NotesAdapter(private val mOnItemClickListener: OnItemClickListener) : PagedListAdapter<Note, BaseViewHolder>(DIFF_CALLBACK) {

    interface OnItemClickListener {
        fun onMenuItemClick(view: View, note: Note)

        fun onItemClick(noteId: Int, noteType: String)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        val view: View
        return when (viewType) {
            R.layout.note_item -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.note_item, parent, false)
                TextNotesViewHolder(view)
            }
            R.layout.image_note_item -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.image_note_item, parent, false)
                ImageNotesViewHolder(view)
            }
            else -> throw IllegalArgumentException(AppConstants.INVALID_VIEW_TYPE)
        }
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    override fun getItemViewType(position: Int): Int {
        val note: Note? = getItem(position)
        return when (note?.noteType) {
            AppConstants.SIMPLE_NOTE,
            AppConstants.CHECKLIST -> R.layout.note_item
            else -> R.layout.image_note_item
        }
    }

    internal inner class TextNotesViewHolder(itemView: View) : BaseViewHolder(itemView), View.OnClickListener {

        init {
            itemView.note_title.setOnClickListener(this)
            itemView.note_description.setOnClickListener(this)
            itemView.note_create_date.setOnClickListener(this)
            itemView.item_menu_button.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            val note = getItem(adapterPosition)!!
            if (view is ImageButton) {
                mOnItemClickListener.onMenuItemClick(view, note)
            } else if (view is TextView) {
                mOnItemClickListener.onItemClick(note.noteId, note.noteType ?: "")
            }
        }

        override fun bind(note: Note?) {
            note?.let {
                if (note.noteType == AppConstants.SIMPLE_NOTE) {
                    itemView.note_description.text = note.description
                } else if (note.noteType == AppConstants.CHECKLIST) {
                    val checklistItems = Gson().fromJson<List<ChecklistItem>>(note.description, object : TypeToken<List<ChecklistItem>>() {
                    }.type)

                    val noOfItems = if (checklistItems.size >= 2) 2 else checklistItems.size
                    itemView.note_description.text = ""
                    for (i in 0 until noOfItems) {
                        itemView.note_description
                            .append(checklistItems!![i].title)
                        if (checklistItems[i].isChecked) {
                            itemView.note_description.append(" " + itemView.context.getString(R.string.checkmark_unicode))
                        }
                        if (i < noOfItems - 1) {
                            itemView.note_description.append("\n")
                        }
                    }
                    if (checklistItems != null && checklistItems.size > 2) {
                        itemView.note_description.append(" ...")
                    }
                }
                itemView.note_title.text = note.noteTitle
                itemView.note_create_date.text = NoteUtils.getFormattedTime(note.dateCreated,
                    System.currentTimeMillis())
            }
        }
    }

    internal inner class ImageNotesViewHolder(itemView: View) : BaseViewHolder(itemView), View.OnClickListener {

        init {
            itemView.tv_note_title.setOnClickListener(this)
            itemView.note_image_view.setOnClickListener(this)
            itemView.tv_note_date_created.setOnClickListener(this)
            itemView.note_menu_button.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            val note = getItem(adapterPosition)!!
            if (view is ImageButton) {
                mOnItemClickListener.onMenuItemClick(view, note)
            } else if (view is TextView || view is ImageView) {
                mOnItemClickListener.onItemClick(note.noteId, note.noteType ?: "")
            }
        }

        override fun bind(note: Note?) {
            note?.let {
                val imageFile = File(note.description!!)
                if (imageFile.exists()) {
                    Glide.with(itemView.context).asDrawable().load(Uri.fromFile(imageFile)).into(itemView.note_image_view)
                }
                itemView.tv_note_title.text = note.noteTitle
                itemView.tv_note_date_created.text = NoteUtils.getFormattedTime(note.dateCreated,
                    System.currentTimeMillis())
            }
        }
    }

    companion object {
        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Note>() {
            override fun areItemsTheSame(old: Note, new: Note): Boolean {
                return old.noteId == new.noteId
            }

            override fun areContentsTheSame(old: Note, new: Note): Boolean {
                return old == new
            }
        }
    }
}
